### ISSUE TITLE
Ansible does not behave like plain Jinja2 with chained `map(attribute='x') | select()`

### ISSUE THREAD

**Issue by cans** [ansible/ansible#24338]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

Ansible Jinja filters used in when clause

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = ... (see below)
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### CONFIGURATION

```[defaults]
roles_path=../../
host_key_checking=false
pipelining=true
```

##### OS / ENVIRONMENT

Ansible 2.3.0 on Debian Jessie, in a virtualenv containing:

ansible (2.3.0.0)
appdirs (1.4.3)
asn1crypto (0.22.0)
backports.shutil-get-terminal-size (1.0.0)
cffi (1.10.0)
cryptography (1.8.1)
decorator (4.0.11)
enum34 (1.1.6)
idna (2.5)
ipaddress (1.0.18)
ipython (5.3.0)
ipython-genutils (0.2.0)
Jinja2 (2.8)
MarkupSafe (1.0)
packaging (16.8)
paramiko (2.1.2)
passlib (1.7.1)
pathlib2 (2.2.1)
pexpect (4.2.1)
pickleshare (0.7.4)
pip (9.0.1)
pkg-resources (0.0.0)
prompt-toolkit (1.0.14)
ptyprocess (0.5.1)
pyasn1 (0.2.3)
pycparser (2.17)
pycrypto (2.6.1)
Pygments (2.2.0)
PyYAML (3.12)
scandir (1.5)
setuptools (35.0.2)
simplegeneric (0.8.1)
six (1.10.0)
traitlets (4.3.2)
wcwidth (0.1.7)
wheel (0.29.0)

##### SUMMARY

In conditionals (and elsewhere, cf debug message in the playbook below), a `map(attribute='x')` applied on a list of `dict` and followed with a `select()` does not seem to behave in the same way as in plain Jinja2.

The issue may lay in the handling of the `Undefined` values in the mapped list, which happens when the `'x'` key not found int one of the dictionary in the list. 

##### STEPS TO REPRODUCE

Run the playbook below with the following command (with ansible.cfg containing the above):

```shell
ANSIBLE_CONFIG=ansible.cfg ansible-playbook -vvvvvv --connection=local issue.yml
```

```yaml
---
- hosts: localhost
  gather_facts: false
  vars:
    list_of_servers:
      - name: first.default.test
        default: true
      - name: second.default.test.realm
        default: true
      - name: third.test.realm
      - name: fourth.test.realm
        default: false
  tasks:
    - name: Show list_of_servers
      debug: var="list_of_servers"
    
    - name: Show map
      debug:
        msg: "{{ list_of_servers | map(attribute='default') | list }}"
    
    - name: Show map + select
      debug:
        msg: "{{ list_of_servers | map(attribute='default') | select() | list }}"
    
    - name: Check list of servers definition for inconsistencies
      fail:
        msg: "Only one default is allowed (you have {{ list_of_servers | map(attribute='default') | select() | list | length }})"
      when: list_of_servers | map(attribute='default') | select() | list | length) > 1
   
```

##### EXPECTED RESULTS

The same as with plain Jinja2:

```python
import jinja2
t2 = jinja2.Template("{{ list_of_servers | map(attribute='default') | select() | list | length }}")
print(t2.render(list_of_servers=[{'name': 'a.b.c', }, {'name': 'a.b.c', 'default': False}, {'name': 'a.b.c', 'default': True}]))
```
Which will output: `1`

##### ACTUAL RESULTS

```ANSIBLE_CONFIG=roles/kerberize/tests/ansible.cfg ansible-playbook -i roles/kerberize/tests/inventory.local --connection=local issue.yml
 ____________________________
< TASK [Show map + select] >
 ----------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

fatal: [localhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'dict object' has no attribute 'default'\n\nT}
	to retry, use: --limit @issue.retry
 ____________
```

**Comment by jctanner** [comment]
@cans this is a closer representation of how ansible uses jinja:

```
#!/usr/bin/env python

from pprint import pprint
from jinja2.environment import Environment
from jinja2.runtime import StrictUndefined

list_of_servers = [
    {
        'name': 'first.default.test',
        'default': True
    },
    {
        'name': 'second.default.test.realm',
        'default': True
    },
    {
        'name': 'third.test.realm'
    },
    {
        'name': 'fourth.test.realm',
        'default': False
    }
]

example = "{{ list_of_servers | map(attribute='default') | select() | list }}"
env = Environment(undefined=StrictUndefined)
tmpl = env.from_string(example)
res = tmpl.render(list_of_servers=list_of_servers)
pprint(res)
```

```
$ ./example.py
Traceback (most recent call last):
  File "./example.py", line 28, in <module>
    res = tmpl.render(list_of_servers=list_of_servers)
  File "/home/jtanner/workspace/github/jctanner/jinja/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/jtanner/workspace/github/jctanner/jinja/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/home/jtanner/workspace/github/jctanner/jinja/jinja2/filters.py", line 769, in do_list
    return list(value)
  File "/home/jtanner/workspace/github/jctanner/jinja/jinja2/filters.py", line 1021, in select_or_reject
    if func(item):
  File "/home/jtanner/workspace/github/jctanner/jinja/jinja2/filters.py", line 1014, in <lambda>
    return seq, lambda item: modfunc(func(transfunc(item)))
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'default'
```

We construct the environment with undefined=StrictUndefined and that is the reason your example template fails with ansible but not in your example jinja code.

**Comment by cans** [comment]
Thanks for the feedback.

I guess you have good reasons to do so, but makes it harder to describe things in a concise way. It is easier to see something's wrong in

```yaml
list_of_servers:
  - name: first.default.test
    default: true
  - name: second.default.test.realm
    default: true
  - name: third.test.realm
  - name: fourth.test.realm
```
than in
```yaml
list_of_servers:
  - name: first.default.test
    default: true
  - name: second.default.test.realm
    default: false
  - name: third.test.realm
    default: false
  - name: fourth.test.realm
    default: true
```
Don't you think ?

I guess a sort of pre-processing step to "fill the gaps" in the former could help (see work-around below). But doing so, with Jinja, is not very easy either. Would a configuration directive to disable Jinja's `StrictUndefined` something realistic ? 

Maybe extending select so one can tell it to locally eval `Undefined` as `False` ? Or even an explicit value:

```
   some_list_of_dicts | map(attribute='x') | select(undefined=value) | list
```

Can we turn this bug into a enhancement / feature request then ? Shall I close this issue and open a new one ?

For the record, here is how I worked around this:

```yaml
- hosts: localhost
  vars:
    list_of_servers:
      - name: first.default.test
        default: true
      - name: second.default.test.realm
        default: true
      - name: third.test.realm
      - name: fourth.test.realm
        default: false
    # Define defaults for items in list_of_servers
    _server_defaults:
      default: false
      # name: considered mandatory so left out
  tasks:
    - name: Show list_of_servers
      debug: var="list_of_servers"

    - name: "Enforce server schema"
      set_fact:
        # Combine the defaults with the data actually given by the "user"
        "{{ _server_defaults | combine(item) }}"
      with_items: "{{ list_of_servers }}"
      register: _list_of_servers
      run_once: true  # Slight optimization.

    # - name: Show *schema enforced* list of servers
    #   debug:
    #     var: _list_of_servers

    - name: Show map + selected
      debug:
        msg: "{{ _list_of_servers.results | map(attribute='ansible_facts._raw_params.default') | select() | list }}"

    - name: Check list of servers definition for inconsistencies
      fail:
        msg: "Only one default is allowed (you have {{ _list_of_servers.results | map(attribute='ansible_facts._raw_params.default') | select() | list | length }})"
      when: _list_of_servers.results | map(attribute='ansible_facts._raw_params.default') | select() | list | length > 1
```

**Comment by jctanner** [comment]
@cans Does this expression work for you and does it give the expected output with ansible?

```
"{{ list_of_servers | map(attribute='default') | select('defined') | select('equalto', True) | list }}"
```

needs_info

**Comment by cans** [comment]
@jctanner: I had not thought of that. Sadly it does not work as what the filter produces is a list of booleans that say whether the 'default' attribute was defined or not. But you loose the actual value of `default`. What I am trying to do is ensure there is only one `default: true`. So I simply count the values equal to `true` (assuming that when undefined `default` defaults to false) making sure there is no more than 1.

The list below:
```yaml
list_of_servers:
- name: A
  default: true
- name: B
  default: false
- name: C
```
if left without any implicit values should be equivalent to:
```yaml
list_of_servers:
- name: A
  default: true
- name: B
  default: false
- name: C
  default: false
```
Which means I would like `map(attribute='default') | select()` to yield `[true, false, false] `.

While `maps(attribute='default') | select('defined')` yields `[true, true, false]`, because the `default` key is defined in the two first dictionaries in the list, not in the third. Hence the result is not what I expect (here I would say that there are too many 'default: true' which is wrong).

Of course, another approach would be for me to implement a filter of my own. But what I am trying to achieve seems fairly basic and reusable for it to be in ansible's core, don't you think?

**Comment by cans** [comment]
@jctanner: What you pointed to me, gave me the following idea:

```
"{{ list_of_servers | map(attribute='default') | reject('undefined') | select() | list }}"
```

Note that it only works because the value I am interested in is a boolean and that I assume `Undefined` is equivalent to `false`. Thus, it may not be a valid solution in many other situations.

**Comment by jctanner** [comment]
@cans I dug in to the error_on_undefined_vars option in ansible.cfg because at first glance it seems like that should have controlled StrictUndefined. However, it's behavior is simply return the original jinja expression and it's been that way since ansible 1.3

I chatted with the rest of the team briefly this morning and right now there is no desire to write or accept a feature that would toggle the undefined=StrictUndefined argument in the environment constructor. That's not to say we _never_ will, but it would probably have to be driven through a proposal process to get any traction.

If this is important to you, please create a proposal. https://github.com/ansible/proposals/blob/master/proposals_process_proposal.md 

If you have any questions, please stop by #ansible-devel on freenode and we will try to help.

**Comment by cans** [comment]
@jctanner: Fair enough. I found several workarounds and the latest we came up with is fairly succinct and "elegant" enough for me (thank you again for pointing me in a better direction). Some of the of the limits of the workarounds are also fairly well pointed out in the above discussion.

Hence, since the issue had not been reported so far (AFAIK) and unless it gets some traction from other users (e.g. with use cases for which it is harder to work around), I'll leave it at that for now.

Thanks
